### PR TITLE
✨ Support value specs for Ports

### LINKED ISSUE #1415: Support value_specs for Ports
/kind feature

**Describe the solution you'd like**

We have a need for using [value_specs](https://docs.openstack.org/heat/zed/template_guide/openstack.html#OS::Neutron::Port-prop-value_specs) for Ports. This is an optional map of extra parameters to send as part of the request.

~~Unfortunately, gophercloud does not seem to [want to support this directly](https://github.com/gophercloud/gophercloud/issues/2180).
However, we can still work around that the same way the terraform provider for openstack has done (see [this](https://github.com/terraform-provider-openstack/terraform-provider-openstack/blob/1db102468904804e2613d02a89d7949474549d71/openstack/types.go#L53-L63) and [this](https://github.com/terraform-provider-openstack/terraform-provider-openstack/blob/1db102468904804e2613d02a89d7949474549d71/openstack/util.go#L24)).~~

Edit: Gophercloud has added support for value_specs in https://github.com/gophercloud/gophercloud/pull/2537. Now we just need to make use of it and expose it to the users of CAPO.

**Anything else you would like to add:**

### PR DESCRIPTION
**Pull request by lentzi90**
**What this PR does / why we need it**:

This adds support for Port [value_specs](https://docs.openstack.org/heat/zed/template_guide/openstack.html#OS::Neutron::Port-prop-value_specs).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1415

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.
2. Related PR in gophercloud: https://github.com/gophercloud/gophercloud/pull/2537

**TODOs**:

- [x] squashed commits
- if necessary:
  - [ ] includes documentation
  - [x] adds unit tests

/hold

### CODE REVIEW

**Comment by lentzi90** [comment]
Marked as ready for review, but please note that we need to bump the gopherclouds version to get support for the value_specs.
To test it I have added a commit that replaces the gophercloud version with the commit where support for value_specs was added.

**Comment by lentzi90** [comment]
Gophercloud version bump is merged and test commit dropped :tada:

File: api/v1alpha6/types.go [hunk 1]
```diff
@@ -145,6 +145,9 @@ type PortOpts struct {
 	// These tags are applied in addition to the instance's tags, which will also be applied to the port.
 	// +listType=set
 	Tags []string `json:"tags,omitempty"`
+
+	// Value specs are extra parameters to include in the request.
+	ValueSpecs *map[string]string `json:"valueSpecs,omitempty"`
```
Comment (mdbooth) [inline api/v1alpha6/types.go:150]: I know we use them elsewhere (just above in Profile, for eg), and CAPI uses them extensively, but use of maps like this [goes against k8s API guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps). We at least shouldn't add more of them.

Please could you replace this with a list with a name as a key element?

I also wonder if we could improve the docstring.

**Review by mdbooth — CHANGES_REQUESTED** [review]

File: api/v1alpha6/types.go [hunk 2]
```diff
@@ -316,3 +321,14 @@ type APIServerLoadBalancer struct {
 	// AllowedCIDRs restrict access to all API-Server listeners to the given address CIDRs.
 	AllowedCIDRs []string `json:"allowedCidrs,omitempty"`
 }
+
+// ValueSpec represents a single value_spec key-value pair.
+type ValueSpec struct {
+	// Name is the name of the key-value pair.
+	// This is just for identifying the pair and will not be sent to the OpenStack API.
```
Comment (mdbooth) [inline api/v1alpha6/types.go:328]: This needs:

```go
  // +required
  // +kubebuilder:validation:Required
```

Although tbh even after reading [the documentation](https://book.kubebuilder.io/reference/markers/crd-validation.html) I don't know what the difference is between those 2, but they always seem to go together 🤷

Comment (mdbooth) [reply api/v1alpha6/types.go:328]: @JoelSpeed I don't suppose you happen to know 👆?

File: api/v1alpha6/types.go [hunk 3]
```diff
@@ -145,6 +145,11 @@ type PortOpts struct {
 	// These tags are applied in addition to the instance's tags, which will also be applied to the port.
 	// +listType=set
 	Tags []string `json:"tags,omitempty"`
+
+	// Value specs are extra parameters to include in the API request with OpenStack.
+	// This is an extension point for the API, so what they do and if they are supported,
+	// depends on the specific OpenStack implementation.
+	ValueSpecs []ValueSpec `json:"valueSpecs,omitempty"`
```
[the remark this answers is not in the document]
Comment (lentzi90) [reply api/v1alpha6/types.go:155]: Thanks, fixed!

File: api/v1alpha6/types.go (hunk 2 above)
Comment (lentzi90) [reply api/v1alpha6/types.go:331]: This does not seem to make any difference. I tried `make generate` with and without and nothing in the CRDs change. All fields are already correctly marked as required. Maybe there is a default somewhere? I see a lot of `+optional` but could nto find a single occurrence of `+required` in the repo :thinking:

Comment (JoelSpeed) [reply api/v1alpha6/types.go:331]: There's no such things as `+required`, but `+kubebuilder:validation:Required` should be added for completeness. Yes, the default is to be required, but someone could, per API group, decide to change the default to optional at some point, in which case, having defined required fields as explicitly required, is safer (and also more obvious that you have actively thought about whether it should be required or optional)

Comment (lentzi90) [reply api/v1alpha6/types.go:331]: Thanks for the explanation! I added `+kubebuilder:validation:Required`.

**Review by mdbooth — APPROVED** [review]

**Review by furkatgofurov7 — COMMENTED** [review]
/lgtm

**Pull request merged**